datasetinsights/estimators/faster_rcnn.py

```diff
@@ -33,44 +36,102 @@ class FasterRCNN(Estimator):
         checkpointer: Model checkpointer callback to save models
         device: model training on device (cpu|cuda)
+        gpu (int): (optional) gpu id on which code will execute
+        rank (int): (optional) rank of process executing code
     """
 
     def __init__(
         self,
         *,
         config,
-        writer,
-        kfp_writer,
-        checkpointer,
-        device,
+        logdir,
         box_score_thresh=0.05,
-        gpu=0,
-        rank=0,
+        data_root=None,
+        no_cuda=None,
+        kfp_metrics_filename=const.DEFAULT_KFP_METRICS_FILENAME,
+        kfp_metrics_dir=const.DEFAULT_KFP_METRICS_DIR,
         **kwargs,
     ):
         """initiate estimator."""
+
         logger.info(f"initializing faster rcnn")
         self.config = config
-        self.device = device
-        self.writer = writer
-        self.kfp_writer = kfp_writer
+
+        if "RANK" in os.environ and "WORLD_SIZE" in os.environ:
+            logger.info(f"found RANK and WORLD_SIZE in environment")
+            self.rank = int(os.environ["RANK"])
+            self.world_size = int(os.environ["WORLD_SIZE"])
+            self.gpu = int(os.environ["LOCAL_RANK"])
+        elif "SLURM_PROCID" in os.environ:
+            logger.info(f"found 'SLURM_PROCID' in environment")
+            self.rank = int(os.environ["SLURM_PROCID"])
+            self.gpu = self.rank % torch.cuda.device_count()
+        else:
+            self.gpu = 0
+            self.rank = 0
+            logger.info("Not using distributed mode")
+            self.distributed = False
+
+        if self.gpu != 0 and self.rank != 0:
+            device_count = torch.cuda.device_count()
+            logger.info(f"device count: {torch.cuda.device_count()}")
+            logger.info(f"world size: {self.world_size}")
+            logger.info(f"gpu: {self.gpu}")
+            logger.info(f"local rank {self.rank}")
+            if device_count == 0:
+                logger.info("No cuda devices found, will not parallelize")
+                self.distributed = False
+            else:
+                if not is_master():
+                    logging.disable(logging.ERROR)
+                self.distributed = True
+                torch.cuda.set_device(self.gpu)
+
+                torch.distributed.init_process_group(
+                    backend="nccl",
+                    init_method="env://",
+                    world_size=self.world_size,
+                    rank=self.rank,
+                )
+                torch.distributed.barrier()
+
+        if torch.cuda.is_available() and not no_cuda:
+            self.device = torch.device("cuda")
+        else:
+            self.device = torch.device("cpu")
+
+        self.writer = SummaryWriter(logdir, write_to_disk=self.distributed)
+
+        self.kfp_writer = KubeflowPipelineWriter(
+            filename=kfp_metrics_filename, filepath=kfp_metrics_dir,
+        )
+
+        self.checkpointer = EstimatorCheckpoint(
```

Comment:
I think it's better to override the instance variable directly, such as `self.checkpointer.distributed = self.distributed`

Reply:
This is under the assumption that the checkpointer has been initialized with is not the case as we discussed